engine/runtime/src/main/scala/org/enso/interpreter/instrument/execution/JobExecutionEngine.scala

```diff
@@ -70,8 +70,29 @@ class JobExecutionEngine(
     runInternal(job, backgroundJobExecutor, backgroundJobsRef)
 
   /** @inheritdoc */
-  override def run[A](job: Job[A]): Future[A] =
+  override def run[A](job: Job[A]): Future[A] = {
+    cancelDuplicateJobs(job)
     runInternal(job, jobExecutor, runningJobsRef)
+  }
+
+  private def cancelDuplicateJobs[A](job: Job[A]): Unit = {
+    job match {
+      case job: UniqueJob[_] =>
+        val allJobs =
+          runningJobsRef.updateAndGet(_.filterNot(_.future.isCancelled))
+        allJobs.foreach { runningJob =>
+          runningJob.job match {
+            case jobRef: UniqueJob[_]
+                if jobRef.getClass == job.getClass && jobRef.key == job.key =>
```

Comment:
I'm wondering if maybe a better choice would be to simply override `equals` (and `hashCode`) in `UniqueJob` instead.